multicluster/cmd/link-gen.go

```diff
@@ -0,0 +1,415 @@
+package cmd
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/linkerd/linkerd2/controller/gen/apis/link/v1alpha3"
+	pkgcmd "github.com/linkerd/linkerd2/pkg/cmd"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/yaml"
+)
+
+type (
+	linkGenOptions struct {
+		namespace                string
+		clusterName              string
+		apiServerAddress         string
+		serviceAccountName       string
+		gatewayName              string
+		gatewayNamespace         string
+		selector                 string
+		remoteDiscoverySelector  string
+		federatedServiceSelector string
+		gatewayAddresses         string
+		gatewayPort              uint32
+		enableGateway            bool
+		output                   string
+	}
+)
+
+func newGenCommand() *cobra.Command {
```

Comment:
I think this would be a lot easier to follow if we broke this into some reasonably small helper functions. Given that other CLIs don't do this, it's not really a blocker...

Reply:
Done in 881194f7fb7d1dc54eac6fc6fb6b1447932558ab, which also removes duplication in the creds generation code.